src/hooks/useWorkletStateMachine/index.ts

```diff
@@ -0,0 +1,180 @@
+ * 5. We open keyboard. It sends KEYBOARD_OPEN action. idle can react to this action
+ * by transitioning into keyboardOpen state
+ * 6. Our state is keyboardOpen. It can react to KEYBOARD_CLOSE, EMOJI_PICKER_OPEN actions
+ * 7. We open emoji picker again. It sends EMOJI_PICKER_OPEN action which transitions our state
+ * into emojiPickerOpen state. Now we react only to EMOJI_PICKER_CLOSE action.
+ * 8. Before rendering the emoji picker, the app hides the keyboard.
+ * It sends KEYBOARD_CLOSE action. But we ignore it since our emojiPickerOpen state can only handle
+ * EMOJI_PICKER_CLOSE action. So we write the logic for handling hiding the keyboard,
+ * but maintaining the offset based on the keyboard state shared value
+ * 9. We close the picker and send EMOJI_PICKER_CLOSE action which transitions us back into keyboardOpen state.
+ *
+ * State machine object example:
+ * const stateMachine = {
+ *   idle: {
+ *       KEYBOARD_OPEN: 'keyboardOpen',
+ *   },
+ *   keyboardOpen: {
+ *       KEYBOARD_CLOSE: 'idle',
+ *       EMOJI_PICKER_OPEN: 'emojiPickerOpen',
+ *   },
+ *   emojiPickerOpen: {
+ *       EMOJI_PICKER_CLOSE: 'keyboardOpen',
+ *   },
+ * }
+ *
+ * Initial state example:
+ * {
+ *     previous: null,
+ *     current: {
+ *         state: 'idle',
+ *         payload: null,
+ *     },
+ * }
+ *
+ * @param stateMachine - a state machine object
+ * @param initialState - the initial state of the state machine
+ * @returns an object containing the current state, a transition function, and a reset function
+ */
+function useWorkletStateMachine<P>(stateMachine: StateMachine, initialState: State<P>) {
+    const currentState = useSharedValue(initialState);
+
+    const log = useCallback((message: string, params?: P | null) => {
+        'worklet';
+
+        if (!DEBUG_MODE) {
+            return;
+        }
+
+        // eslint-disable-next-line @typescript-eslint/unbound-method, @typescript-eslint/restrict-template-expressions
+        runOnJS(client)(`[StateMachine] ${message}. Params: ${JSON.stringify(params)}`);
+    }, []);
+
+    const transitionWorklet = useCallback(
+        (action: ActionWithPayload<P>) => {
+            'worklet';
+
+            if (!action) {
+                throw new Error('state machine action is required');
+            }
+
+            const state = currentState.get();
+
+            log(`Current STATE: ${state.current.state}`);
+            log(`Next ACTION: ${action.type}`, action.payload);
+
+            const nextMachine = stateMachine[state.current.state];
+
+            if (!nextMachine) {
+                log(`No next machine found for state: ${state.current.state}`);
+                return;
+            }
+
+            const nextState = nextMachine[action.type];
+
+            if (!nextState) {
+                log(`No next state found for action: ${action.type}`);
+                return;
+            }
+
+            let nextPayload;
```

Comment:
This can be simplified:

```ts
// Merge the new payload with the previous payload
const nextPayload = {
    ...state.current.payload,
    ...action.payload,
};
```

This works fine even if `action.payload` is undefined. If you're keen to avoid the shallow copy of `state.current.payload`, and we're ok with mutating it (which, I assume we are because the mutation would only happen outside this scope when we call `.set`), then you could use `Object.assign`:

```ts
// Merge the new payload with the previous payload
const nextPayload = Object.assign(state.current.payload ?? {}, action.payload);
```

I assume that since these payloads are small the spread would be preferred.

*Edit:* The spread seems to cause some type errors, so maybe use `Object.assign`?

Reply:
Hmm, I don't know. It doesn't look clearer to me, after removing comments it became a bit messy
Maybe you would like this one?

```
            // we save previous payload
            let nextPayload = state.current.payload;

            if (typeof action.payload !== 'undefined') {
              // we merge previous payload with the new payload
              nextPayload = Object.assign({}, state.current.payload, action.payload);
            }
```

Reply:
How about this?

```ts
import fastMerge from 'expensify-common/dist/fastMerge';

...

const nextPayload = typeof action.payload === 'undefined'
    ? action.payload
    : fastMerge(state.current.payload, action.payload);
```

Reply:
Ultimately this probably doesn't matter much. I just thought one statement would be cleaner than a let declaration and if/else. `fastMerge` seems like a bonus, but idk if it's faster than spread in the case when a shallow copy is good enough (I doubt it)